File: Data Structures & Algorithms/palindrome-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        if(head == null || head.next == null) return true;
        int size = 0;
        ListNode current = head;
        while(current != null){
            size++;
            current = current.next;
        }

        int[] arr = new int[size];
        current = head;
        for(int i = 0;i<arr.length;i++){
            arr[i] = current.val;
            current = current.next;
        }

        int l = 0;
        int r = arr.length -1;

        while(l<r){
            if(arr[l] != arr[r]){
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
}

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None

        new = {}
        current = head

        while current :
            new[current] = Node(current.val)
            current = current.next

        current = head

        while current:
            copy = new[current]

            copy.next = new.get(current.next)
            copy.random = new.get(current.random)

            current = current.next

        return new[head]
